engine/crates/validation/src/validate/schema_definition.rs

```diff
@@ -0,0 +1,104 @@
+    if let Some(previous_def) = ctx.schema_definition.take() {
+        let labels = vec![
+            miette::LabeledSpan::new_with_span(
+                Some("Previous definition".to_owned()),
+                (
+                    ctx.miette_pos(previous_def.pos),
+                    miette::SourceOffset::from("schema".len()),
+                ),
+            ),
+            miette::LabeledSpan::new_with_span(
+                Some("Second definition".to_owned()),
+                (ctx.miette_pos(def.pos), miette::SourceOffset::from("schema".len())),
+            ),
+        ];
+
+        ctx.push_error(miette::miette! {
+            labels = labels,
+            "Duplicate schema definition",
+        });
+    }
+
+    ctx.schema_definition = Some(SchemaDefinition {
+        pos: def.pos,
+        query: def.node.query.as_ref().unwrap().node.as_str(),
+        mutation: def.node.mutation.as_ref().map(|node| node.node.as_str()),
+        subscription: def.node.subscription.as_ref().map(|node| node.node.as_str()),
+    });
+}
+
+pub(crate) fn validate_schema_definition_references(ctx: &mut Context<'_>) {
+    let Some(def) = ctx.schema_definition.as_ref() else {
+        return;
+    };
+    let pos = def.pos;
+
+    let names = [
+        (Some(def.query), "Query"),
+        (def.mutation, "Mutation"),
+        (def.subscription, "Subscription"),
+    ];
+
+    for idx in 0..(names.len()) {
+        let name = &names[idx];
+        let name = name.0.unwrap_or(name.1);
+        for other_name in names[idx..].iter().skip(1) {
+            let other_name = other_name.0.unwrap_or(other_name.1);
+            if name == other_name {
+                ctx.push_error(miette::miette!("Typed used twice in schema definition: `{name}`"));
+            }
+        }
+    }
+
+    for (actual, default) in &names {
+        let Some(actual) = actual else { continue };
+
+        if actual == default {
+            continue;
+        }
+
+        match ctx.definition_names.get(actual) {
+            None => {
+                let labels = vec![miette::LabeledSpan::new_with_span(
+                    None,
+                    miette::SourceSpan::new(ctx.miette_pos(pos), "schema".len().into()),
+                )];
+                ctx.push_error(miette::miette!(
+                    labels = labels,
+                    "Cannot set schema {} root to unknown type type `{actual}`",
+                    default.to_lowercase()
+                ));
+            }
+            // http://spec.graphql.org/draft/#sec-Root-Operation-Types
+            Some(referenced) => match referenced.node.kind {
+                ast::TypeKind::Object(_) => (),
+                _ => {
+                    let type_name = referenced.node.name.node.as_str();
+                    ctx.push_error(miette::miette!(
+                        "Query root type must be an Object type, it cannot be set to {type_name}"
```

Comment:
Should this error message say "Query/Mutation/Subscription root type" rather than just `Query root type`?

Reply:
It's one of the not-so-great errors I copied from the ecosystem. I agree we should do better here, I'll adjust.

Reply:
Done and pushed